dart/lib/src/protocol/contexts.dart

```diff
@@ -11,107 +13,169 @@ import 'runtime.dart';
-        for (final runtime in runtimes) {
-          if (runtime != null) {
-            var key = runtime.key ?? runtime.name.toLowerCase();
-
-            if (json.containsKey(key)) {
-              var k = 0;
-              while (json.containsKey(key)) {
-                key = '$key$k';
-                k++;
+    forEach((key, value) {
+      if (value == null) return;
+      switch (key) {
+        case Device.type:
+          Map<String, dynamic> deviceMap;
+          if ((deviceMap = device.toJson()).isNotEmpty) {
+            json[Device.type] = deviceMap;
+          }
+          break;
+        case OperatingSystem.type:
+          Map<String, dynamic> osMap;
+          if ((osMap = operatingSystem.toJson()).isNotEmpty) {
+            json[OperatingSystem.type] = osMap;
+          }
+          break;
+
+        case App.type:
+          Map<String, dynamic> appMap;
+          if ((appMap = app.toJson()).isNotEmpty) {
+            json[App.type] = appMap;
+          }
+          break;
+
+        case Browser.type:
+          Map<String, dynamic> browserMap;
+          if ((browserMap = browser.toJson()).isNotEmpty) {
+            json[Browser.type] = browserMap;
+          }
+          break;
+
+        case Gpu.type:
+          Map<String, dynamic> gpuMap;
+          if ((gpuMap = gpu.toJson()).isNotEmpty) {
+            json[Gpu.type] = gpuMap;
+          }
+          break;
+
+        case Runtime.listType:
+          if (runtimes != null) {
+            if (runtimes.length == 1) {
+              final runtime = runtimes[0];
+              if (runtime != null) {
+                final key = runtime.key ?? Runtime.type;
+                json[key] = runtime.toJson();
+              }
+            } else if (runtimes.length > 1) {
+              for (final runtime in runtimes) {
+                if (runtime != null) {
+                  var key = runtime.key ?? runtime.name.toLowerCase();
+
+                  if (json.containsKey(key)) {
+                    var k = 0;
+                    while (json.containsKey(key)) {
+                      key = '$key$k';
+                      k++;
+                    }
+                  }
+
+                  json[key] = runtime.toJson()
+                    ..addAll(<String, String>{'type': Runtime.type});
+                }
               }
             }
-
-            json[key] = runtime.toJson()
-              ..addAll(<String, String>{'type': 'runtime'});
           }
-        }
+
+          break;
+
+        default:
+          json[key] = value;
       }
-    }
+    });
 
     return json;
   }
+
+  Contexts clone() {
+    final copy = Contexts(
+      device: device?.clone(),
+      operatingSystem: operatingSystem?.clone(),
+      app: app?.clone(),
+      browser: browser?.clone(),
+      gpu: gpu?.clone(),
+      runtimes: runtimes.map((runtime) => runtime.clone()).toList(),
+    )..addEntries(
+        entries.where((element) => !_defaultFields.contains(element.key)),
+      );
```

Comment:
is this case covered by an unit test?

Reply:
is it enough? https://github.com/getsentry/sentry-dart/blob/99027322f429d6884cf7ec9a85a2053426ffef49/dart/test/contexts_test.dart#L127-L128